### ISSUE TITLE
How are anchors updated during training?

### ISSUE THREAD

**Issue by AlessandroMondin** [ultralytics/yolov5#10201]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I am trying to understand the loss function and I noticed that the anchors are only defined in the ComputeLoss class 
with `self.anchors = m.anchors` ([here](https://github.com/ultralytics/yolov5/blob/master/utils/loss.py#L118)) and I could'n find nor in this class nor in the train.py anything related to updating such anchors.

My question is, where are anchors updated during training?

**Comment by glenn-jocher** [comment]
@AlessandroMondin 👋 Hello! Thanks for asking about model anchors. YOLOv5 🚀 uses a new [Ultralytics](https://ultralytics.com/) algorithm called **AutoAnchor** for anchor verification and generation before training starts.

Autoanchor will analyse your anchors against your dataset and training settings (like `--img-size`), and will adjust your anchors as necessary if it determines the original anchors are a poor fit, or if an **anchor count** was specified in your model.yaml rather than **anchor values**, i.e. 
```yaml
# Specify anchor count (per layer)
anchors: 3

# --OR-- Specify anchor values manually
anchors:
  - [10,13, 16,30, 33,23]  # P3/8
  - [30,61, 62,45, 59,119]  # P4/16
  - [116,90, 156,198, 373,326]  # P5/32
```

When generating new anchors, autoanchor first applies a kmeans function against your dataset labels (scaled to your training `--img-size`), and uses kmeans centroids as initial conditions for a [Genetic Evolution](https://en.wikipedia.org/wiki/Genetic_algorithm) (GE) algorithm. The GE algorithm will evolve all anchors for 1000 generations under default settings, using CIoU loss (same regression loss used during training) combined with [Best Possible Recall](https://en.wikipedia.org/wiki/Precision_and_recall) (BPR) as its fitness function.

Notebook example: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

<img width="1200" alt="Screenshot 2022-03-24 at 10 58 28" src="https://user-images.githubusercontent.com/26833433/159891298-8b15f80b-f5b0-4fa9-8824-7380bae32a76.png">

**No action is required on your part** to use autoanchor. If you would like to force manual anchors for any reason, you can skip autoanchor with the `--noautoanchor` flag:
```shell
python train.py --noautoanchor
```

For more details on AutoAnchor see:
https://github.com/ultralytics/yolov5/blob/master/utils/autoanchor.py

Good luck 🍀 and let us know if you have any other questions!

**Comment by AlessandroMondin** [comment]
Hi Glen! Thanks for sharing!
Therefore after the initial research for best anchors with auto-anchors, these are fixed during yolo training and are not treated as yolo parameters (modified with back-propagation?

**Comment by glenn-jocher** [comment]
@AlessandroMondin yes that's correct, anchors are not modified once training starts, they do not have a gradient.

**Comment by AlessandroMondin** [comment]
Thanks a lot Glen! Doubts solved 👍

**Comment by glenn-jocher** [comment]
@AlessandroMondin I tried making anchors learnable parameters once, but did not get good results this way.

**Comment by bot66** [comment]
@glenn-jocher Hi Glenn, I wonder `yolov5*.yaml` default anchor value is based on the training image size of 640, so if I set the training image size to 1280, does it make any sense if I scale all the anchor width and height to 2 times larger? 
Or even bigger image size, say:5120, then scaling the anchors is a must since the `build_targets` will reject anchors that are smaller than gt bbox after multiplying the `anchor_t` which is 4.

I don't know if this is correct.

**Comment by Yosu26** [comment]
@glenn-jocher Hi Glenn, Is this autoanchor a good thing or not?

**Comment by glenn-jocher** [comment]
@Yosu26 Hey there! 🚀 AutoAnchor is definitely a good thing. It helps ensure that the anchors are well-suited to your specific dataset, potentially improving model accuracy and performance right from the start of training. It automatically adjusts if it detects that your current anchors are not an optimal fit, without requiring manual recalibration. So, yes, it's designed to make your life easier and your model more effective! 😊
